src/storage-controller/src/collection_mgmt.rs

```diff
@@ -1439,46 +1435,76 @@ where
 
     let mut deletions = vec![];
 
-    for (row, diff) in rows {
+    let latest_row_per_key_ref = &mut latest_row_per_key;
+
+    let mut handle_row = move |row: Row, diff| {
```

Comment:
This is something I'm unsure about. 

I followed Nikhil's suggestion on creating a closure https://github.com/MaterializeInc/materialize/pull/29795#discussion_r1781698384 but when I tried a non-`move` closure w/ `&Row`, I got the following compiler error:
```
error[E0277]: `dyn for<'a, 'b> std::ops::Fn(&'a [mz_repr::Datum<'b>]) -> mz_repr::GlobalId + std::marker::Send` cannot be shared between threads safely
    --> src/storage-controller/src/collection_mgmt.rs:1051:18
     |
1051 |       let handle = mz_ore::task::spawn(
     |  __________________^
1052 | |         || format!("CollectionManager-append_only_write_task-{id}"),
1053 | |         async move {
1054 | |             if !read_only {
...    |
1198 | |         },
1199 | |     );
     | |_____^ `dyn for<'a, 'b> std::ops::Fn(&'a [mz_repr::Datum<'b>]) -> mz_repr::GlobalId + std::marker::Send` cannot be shared between threads safely
     |
     = help: the trait `std::marker::Sync` is not implemented for `dyn for<'a, 'b> std::ops::Fn(&'a [mz_repr::Datum<'b>]) -> mz_repr::GlobalId + std::marker::Send`, which is required by `{async block@src/storage-controller/src/collection_mgmt.rs:1053:9: 1198:10}: std::marker::Send`
     = note: required for `std::ptr::Unique<dyn for<'a, 'b> std::ops::Fn(&'a [mz_repr::Datum<'b>]) -> mz_repr::GlobalId + std::marker::Send>` to implement `std::marker::Sync`
note: required because it appears within the type `std::boxed::Box<dyn for<'a, 'b> std::ops::Fn(&'a [mz_repr::Datum<'b>]) -> mz_repr::GlobalId + std::marker::Send>`
    --> /Users/sangjunbak/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/alloc/src/boxed.rs:237:12
     |
237  | pub struct Box<
     |            ^^^
     = note: required for `&std::boxed::Box<dyn for<'a, 'b> std::ops::Fn(&'a [mz_repr::Datum<'b>]) -> mz_repr::GlobalId + std::marker::Send>` to implement `std::marker::Send`
note: required because it's used within this closure
    --> src/storage-controller/src/collection_mgmt.rs:1440:26
     |
1440 |     let mut handle_row = |row: Row, diff| {
     |                          ^^^^^^^^^^^^^^^^
note: required because it's used within this `async` fn body
    --> src/storage-controller/src/collection_mgmt.rs:1413:1
     |
1413 | / {
1414 | |     let upper = write_handle.fetch_recent_upper().await.clone();
1415 | |
1416 | |     let mut rows = match upper.as_option() {
...    |
1557 | |         .collect()
1558 | | }
     | |_^
note: required because it's used within this `async` fn body
    --> src/storage-controller/src/collection_mgmt.rs:1220:1
     |
1220 | / {
1221 | |     tracing::info!(%id, ?introspection_type, "preparing append only introspection collection for writes");
1222 | |
1223 | |     match introspection_type {
...    |
1298 | |     }
1299 | | }
     | |_^
note: required because it's used within this `async` block
    --> src/storage-controller/src/collection_mgmt.rs:1053:9
     |
1053 | /         async move {
1054 | |             if !read_only {
1055 | |                 if let Some(AppendOnlyIntrospectionConfig {
1056 | |                     introspection_type,
...    |
1197 | |             info!("write_task-{id} ending");
1198 | |         },
     | |_________^
note: required by a bound in `mz_ore::task::spawn`
    --> /Users/sangjunbak/materialize/src/ore/src/task.rs:221:19
     |
217  | pub fn spawn<Fut, Name, NameClosure>(nc: NameClosure, future: Fut) -> JoinHandle<Fut::Output>
     |        ----- required by a bound in this function
...
221  |     Fut: Future + Send + 'static,
     |                   ^^^^ required by this bound in `spawn`


```
Hard to decipher but I figured it's because async closures need ownership, so I converted it into a `move` closure and initialized `latest_row_per_key` as a mut reference such that its ownership isn't transferred. Let me know if there's a better way!

Reply:
Yeah the borrow errors get tricky here! First note is that using `&Row` here with a move closure should still be fine. This compiles for me:

```diff
diff --git a/src/storage-controller/src/collection_mgmt.rs b/src/storage-controller/src/collection_mgmt.rs
index ec111d8f58..e6096c0eaa 100644
--- a/src/storage-controller/src/collection_mgmt.rs
+++ b/src/storage-controller/src/collection_mgmt.rs
@@ -1437,7 +1437,7 @@ where
 
     let latest_row_per_key_ref = &mut latest_row_per_key;
 
-    let mut handle_row = move |row: Row, diff| {
+    let mut handle_row = move |row: &Row, diff| {
         let datums = row.unpack();
         let key = (status_history_desc.extract_key)(&datums);
         let timestamp = (status_history_desc.extract_time)(&datums);
@@ -1455,7 +1455,7 @@ where
                 (latest_row_per_key_ref).insert(key.clone(), (timestamp, row.clone()));
             }
         };
-        (row, key, timestamp)
+        (key, timestamp)
     };
 
     match status_history_desc.retention_policy {
@@ -1467,7 +1467,7 @@ where
             > = BTreeMap::new();
 
             for (row, diff) in rows {
-                let (row, key, timestamp) = handle_row(row, diff);
+                let (key, timestamp) = handle_row(&row, diff);
 
                 // Duplicate rows ARE possible if many status changes happen in VERY quick succession,
                 // so we handle duplicated rows separately.
@@ -1501,7 +1501,7 @@ where
 
             // Mark any row outside the retention window for deletion
             for (row, diff) in rows {
-                let (row, _, timestamp) = handle_row(row, diff);
+                let (_, timestamp) = handle_row(&row, diff);
 
                 if *timestamp < keep_since {
                     deletions.push(row);
```

And then here's an idiomatic way of forcing a `move` closure to capture a variable by reference:

```rust
    let mut handle_row = {
        let latest_row_per_key = &mut latest_row_per_key;
        move |row: &Row, diff| {
            let datums = row.unpack();
            let key = (status_history_desc.extract_key)(&datums);
            let timestamp = (status_history_desc.extract_time)(&datums);

            assert!(
                diff > 0,
                "only know how to operate over consolidated data with diffs > 0, \
                    found diff {diff} for object {key:?} in {introspection_type:?}",
            );

            // Keep track of the timestamp of the latest row per key.
            match latest_row_per_key.get(&key) {
                Some(existing) if &existing.0 > &timestamp => {}
                _ => {
                    latest_row_per_key.insert(key.clone(), (timestamp, row.clone()));
                }
            };
            (key, timestamp)
        }
    };
```

In Rust, it's generally idiomatic to use scoping/shadowing to construct references to values for closures, rather than having to come up with new names like `xxx_ref`.

Here's a patch that puts it all together (hard to read because of whitespace issues though):

```diff
diff --git a/src/storage-controller/src/collection_mgmt.rs b/src/storage-controller/src/collection_mgmt.rs
index ec111d8f58..13c94f2105 100644
--- a/src/storage-controller/src/collection_mgmt.rs
+++ b/src/storage-controller/src/collection_mgmt.rs
@@ -1435,27 +1435,28 @@ where
 
     let mut deletions = vec![];
 
-    let latest_row_per_key_ref = &mut latest_row_per_key;
+    let mut handle_row = {
+        let latest_row_per_key = &mut latest_row_per_key;
+        move |row: &Row, diff| {
+            let datums = row.unpack();
+            let key = (status_history_desc.extract_key)(&datums);
+            let timestamp = (status_history_desc.extract_time)(&datums);
 
-    let mut handle_row = move |row: Row, diff| {
-        let datums = row.unpack();
-        let key = (status_history_desc.extract_key)(&datums);
-        let timestamp = (status_history_desc.extract_time)(&datums);
+            assert!(
+                diff > 0,
+                "only know how to operate over consolidated data with diffs > 0, \
+                    found diff {diff} for object {key:?} in {introspection_type:?}",
+            );
 
-        assert!(
-            diff > 0,
-            "only know how to operate over consolidated data with diffs > 0, \
-                found diff {diff} for object {key:?} in {introspection_type:?}",
-        );
-
-        // Keep track of the timestamp of the latest row per key.
-        match (latest_row_per_key_ref).get(&key) {
-            Some(existing) if &existing.0 > &timestamp => {}
-            _ => {
-                (latest_row_per_key_ref).insert(key.clone(), (timestamp, row.clone()));
-            }
-        };
-        (row, key, timestamp)
+            // Keep track of the timestamp of the latest row per key.
+            match latest_row_per_key.get(&key) {
+                Some(existing) if &existing.0 > &timestamp => {}
+                _ => {
+                    latest_row_per_key.insert(key.clone(), (timestamp, row.clone()));
+                }
+            };
+            (key, timestamp)
+        }
     };
 
     match status_history_desc.retention_policy {
@@ -1467,7 +1468,7 @@ where
             > = BTreeMap::new();
 
             for (row, diff) in rows {
-                let (row, key, timestamp) = handle_row(row, diff);
+                let (key, timestamp) = handle_row(&row, diff);
 
                 // Duplicate rows ARE possible if many status changes happen in VERY quick succession,
                 // so we handle duplicated rows separately.
@@ -1501,7 +1502,7 @@ where
 
             // Mark any row outside the retention window for deletion
             for (row, diff) in rows {
-                let (row, _, timestamp) = handle_row(row, diff);
+                let (_, timestamp) = handle_row(&row, diff);
 
                 if *timestamp < keep_since {
                     deletions.push(row);
```

Reply:
> First note is that using &Row here with a move closure should still be fine.
Oh nice, so I guess ownership's still transferred but we can still have the closure accept a reference? 

And super clean!

Reply:
Yeah, exactly. The way it works is that `move` controls whether the closure takes ownership or not of variables referenced in the closure body. It doesn't have any effect on the closure's parameters.